plugins/inputs/t128_tank/reader.go

```diff
@@ -0,0 +1,354 @@
+package t128_tank
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"math"
+	"os"
+	"os/exec"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	endIndexUint   = math.MaxUint64
+	endIndexStr    = "end"
+	startIndexUint = 0
+	startIndexStr  = "0"
+)
+
+var (
+	StartIndex         = index{value: startIndexUint}
+	EndIndex           = index{value: endIndexUint}
+	boundaryFaultRegex = regexp.MustCompile(`^Boundary Check fault. first available sequence number is (\d+), .*$`)
+	indexRegex         = regexp.MustCompile(`seq=(\d+?):(.*)$`)
+)
+
+type boundaryFault struct {
+	nextAvailableIndex index
+}
+
+func (bf *boundaryFault) Error() string {
+	return fmt.Sprintf("next available index is %d", bf.nextAvailableIndex.value)
+}
+
+// InfluxLine
+type IndexedMessage struct {
+	Message   []byte
+	Index     index
+	Timestamp int
+}
+
+type index struct {
+	value uint64
+}
+
+func newIndex(data string) (index, error) {
+	if string(data) == endIndexStr {
+		return EndIndex, nil
+	}
+	indexVal, err := strconv.ParseUint(data, 10, 64)
+	if err != nil {
+		return StartIndex, fmt.Errorf("unable to parse %s as integer: %s", data, err)
+	}
+	return index{value: indexVal}, nil
+}
+
+func (i index) string() string {
+	if i.value == endIndexUint {
+		return endIndexStr
+	}
+	return fmt.Sprintf("%d", i.value)
+}
+
+func (i index) next() index {
+	if i.value == startIndexUint || i.value == endIndexUint {
+		return i
+	}
+	return index{value: i.value + 1}
+}
+
+type CommandContext = func(ctx context.Context, name string, arg ...string) *exec.Cmd
+
+type Reader struct {
+	topic          string
+	lastSavedIndex uint64
+	// used to send events from the read routine to the send routine
+	sendChan chan IndexedMessage
+	// the target address for the TANK instance
+	tankAddress string
+	// the target port for the TANK instance
+	tankPort int
+	// time to wait before restarting the tank read command
+	restartDelay time.Duration
+	// used to notify the main routine that the read routine finished with an error
+	readDone chan error
+	// used to mock tank read subprocess in testing
+	tankReadCmdCtx CommandContext
+	// path to the index file
+	indexPath string
+	// default index in case reading to/from the index file fails
+	defaultIndex index
+}
+
+func NewReader(tankAddress string, tankPort int, topic string, indexPath string, defaultIndex index) *Reader {
+	return &Reader{
+		topic:          topic,
+		tankReadCmdCtx: exec.CommandContext,
+		sendChan:       make(chan IndexedMessage),
+		readDone:       make(chan error),
+		restartDelay:   5 * time.Second,
+		tankAddress:    tankAddress,
+		tankPort:       tankPort,
+		indexPath:      indexPath,
+		defaultIndex:   defaultIndex,
+	}
+}
+
+func (r *Reader) withTankReadCommandContext(tankReadCmdCtx CommandContext) *Reader {
+	r.tankReadCmdCtx = tankReadCmdCtx
+	return r
+}
+
+func (r *Reader) Run(mainCtx context.Context, stopchan chan struct{}) {
```

Comment:
`stopchan` is unused

Reply:
I forgot to add the case! Will add that!